### ISSUE TITLE
403 Forbidden for gitlab_project on 3.16.1

### ISSUE THREAD

**Issue by thenom** [gitlabhq/terraform-provider-gitlab#1208]
### GitLab Provider version

gitlabhq/gitlab v3.16.1

### GitLab version

GitLab Enterprise Edition 15.3.0-pre 15a9431643b

### Terraform version

1.2

### Relevant Terraform Configuration

```hcl
# module.aws-glue-tables-repo.gitlab_project.project will be updated in-place
  ~ resource "gitlab_project" "project" {
      ~ http_url_to_repo                                 = "https://gitlab.com/company/data-architecture/data-platform/terraform-modules/aws-glue-tables.git" -> (known after apply)
        id                                               = "123456789"
        name                                             = "aws-glue-tables"
      ~ path_with_namespace                              = "company/data-architecture/data-platform/terraform-modules/aws-glue-tables" -> (known after apply)
      ~ ssh_url_to_repo                                  = "git@gitlab.com:company/data-architecture/data-platform/terraform-modules/aws-glue-tables.git" -> (known after apply)
        tags                                             = []
      ~ web_url                                          = "https://gitlab.com/company/data-architecture/data-platform/terraform-modules/aws-glue-tables" -> (known after apply)
        # (57 unchanged attributes hidden)
        # (2 unchanged blocks hidden)
    }
```

### Relevant log output

```shell
│ Error: PUT https://gitlab.com/api/v4/projects/123456789: 403 {message: 403 Forbidden}
│ 
│   with module.aws-glue-job-repo.gitlab_project.project,
│   on .terraform/modules/aws-glue-job-repo/modules/project/main.tf line 1, in resource "gitlab_project" "project":
│    1: resource "gitlab_project" "project" {
```

### Description

Hi,

We have recently updated our terraform to use the new protected environment required_approval_count but when it goes to apply any changes on this version we get the error above for every gitlab_project resource.  We also seem to be picking up changes to those 4 read only properties (http_url_to_repo, path_with_namespace, ssh_url_to_repo and web_url) everytime too but these have been more of an irritant than a problem and has always been there before this upgrade.

To access the API we have the tokens stored in SSM which we retrieve in terraform and then pass to the provider which has always worked fine on the previous provider we used (3.14.0).
```
provider "gitlab" {
  alias = "gitlab-data"
  token = data.aws_ssm_parameter.gitlab_token_data.value
}

data "aws_ssm_parameter" "gitlab_token_data" {
  name = "/gitlab-group-tokens/data"
}
```

To confirm this and to make sure i didnt mess anything up i reverted the versions back and removed the added required_approval_count from the protected environments and it is back working again.

Thanks in advance,
Simon

**Comment by timofurrer** [comment]
@thenom 

> We also seem to be picking up changes to those 4 read only properties (http_url_to_repo, path_with_namespace, ssh_url_to_repo and web_url) everytime too

There is a `CustomizeDiff` function on these particular attributes - they are marked as *computed* if either the `namespace_id` or the `path` changed. Was that somehow the case (what does your state say)? 

Can you run the `terraform apply` with debug logs (setting `TF_LOG=debug`) and posting the output in a gist? And make sure to redact any sensitive information. And very the following things:

* are you really using the token you are expecting to use?
* What is the request body of the `PUT` API request for which you get the `403` - does it make sense to you?

**Comment by thenom** [comment]
@timofurrer Thats for the response, i will look into getting this info out with the debug log.  tbh we have only recently imported our existing Gitlab repos into terraform and have always seen this on every run.  I will also double check the namespace_id and path for computed values but as far as a i know they are static.

With regards to the token, this SSM param and provider setup has not changed between the 2 different branches of our so pretty sure it is still correct but hopefully the debug log should give me an idea.

**Comment by thenom** [comment]
Ok, looks like it is using the correct token but just investigating what this is actually doing:
```
-----------------------------------------------------: timestamp=2022-08-11T08:35:06.892Z
2022-08-11T08:35:06.897Z [ERROR] vertex "module.aws-glue-job-repo.gitlab_project.project" error: PUT https://gitlab.com/api/v4/projects/123456798: 403 {message: 403 Forbidden}
2022-08-11T08:35:06.900Z [INFO]  provider.terraform-provider-gitlab_v3.16.1: 2022/08/11 08:35:06 [DEBUG] GitLab API Request Details:
```
Not sure it makes much difference and probably separate processes writing these log entries out but requests directly beneath the 403's above are not the associated ones so had to search for this:
```
---[ REQUEST ]---------------------------------------
PUT /api/v4/projects/123456798 HTTP/1.1
Host: gitlab.com
User-Agent: Terraform/1.2.6 (+https://www.terraform.io/) Terraform-Plugin-SDK/2.10.1 terraform-provider-gitlab/3.16.1
Content-Length: 23
Accept: application/json
Authorization: Bearer <correct-token>
Content-Type: application/json
Accept-Encoding: gzip
{
 "name": "aws-glue-job"
}
```

**Comment by thenom** [comment]
Same issue with curl using the details from above:
```shell
$ curl -XPUT https://gitlab.com/api/v4/projects/123456798 -H "Authorization: Bearer <correct-token>" -H "Content-Type: application/json" -d '{"name": "aws-glue-job"}'
{"message":"403 Forbidden"}
```
Obviously gitlab isnt changing behind the scenes so re-running this without the change for this module to see the difference in the calls or even if this call is occurring as it seems that this is just trying to set the name of the repo.

**Comment by thenom** [comment]
Same problem on 3.16.0, cant test this on any earlier as protected environment `required_approval_count` came in on 3.16.0.

Removed the required approval and used the previous provider of 3.15.1 and i now have no issue.  The `PUT /api/v4/projects/123456798` also doesnt seem to occur in the apply debug logs either.  So something in 3.16.0 is triggering this `PUT`

Still digging but not even sure yet if the issue is related to the `required_approval_count` yet or something else that came in on 3.16.0.

**Comment by thenom** [comment]
The above still occurs when there is no protected environment resource on a project so it doesnt seem to be that.

Went back to simple basics with:
```
provider "aws" {
  region = "eu-west-1"
}

provider "gitlab" {
  alias = "gitlab-data-architecture"
  token = data.aws_ssm_parameter.gitlab_token_data_architecture.value
}

terraform {
  backend "http" {}

  required_providers {
    gitlab = {
      source = "gitlabhq/gitlab"
    }
  }
}

data "aws_ssm_parameter" "gitlab_token_data_architecture" {
  name = "/gitlab-group-tokens/data-architecture"
}

resource "gitlab_project" "project" {
  name        = "simple-test-module"
  description = "To test the latest TF module"

  path                   = ""
  default_branch         = "main"
  merge_requests_enabled = true

  namespace_id = "123456"

  visibility_level = "private"

  provider = gitlab.gitlab-data-architecture
}
```
still getting the read only changes on second run:
```
  # gitlab_project.project will be updated in-place
  ~ resource "gitlab_project" "project" {
      ~ http_url_to_repo                                 = "https://gitlab.com/company/data-architecture/data-platform/terraform-modules/simple-test-module.git" -> (known after apply)
        id                                               = "0987654321"
        name                                             = "simple-test-module"
      ~ path_with_namespace                              = "company/data-architecture/data-platform/terraform-modules/simple-test-module" -> (known after apply)
      ~ ssh_url_to_repo                                  = "git@gitlab.com:company/data-architecture/data-platform/terraform-modules/simple-test-module.git" -> (known after apply)
        tags                                             = []
      ~ web_url                                          = "https://gitlab.com/company/data-architecture/data-platform/terraform-modules/simple-test-module" -> (known after apply)
        # (56 unchanged attributes hidden)
        # (2 unchanged blocks hidden)
    }
```
and on apply i get the current issue so it seems to be this that is triggering the `PUT` but have no idea why this update is occuring:
```
2022-08-11T13:48:11.981Z [INFO]  provider.terraform-provider-gitlab_v3.16.1: 2022/08/11 13:48:11 [DEBUG] GitLab API Request Details:
---[ REQUEST ]---------------------------------------
PUT /api/v4/projects/0987654321 HTTP/1.1
Host: gitlab.com
User-Agent: Terraform/1.2.6 (+https://www.terraform.io/) Terraform-Plugin-SDK/2.10.1 terraform-provider-gitlab/3.16.1
Content-Length: 29
Accept: application/json
Authorization: Bearer <correct-token>
Content-Type: application/json
Accept-Encoding: gzip
{
 "name": "simple-test-module"
}
```

**Comment by thenom** [comment]
Ok, found it...
```
  path = ""
```
is the problem.  We have this as a variable in our module that defaults to "" which i replicated in my simple project.  When i removed the attributes the changes stop and with that, the `PUT` stops.

No-idea why this causes a change or why that triggers a forbidden `PUT` but definitely doesn't seem right.

**Comment by timofurrer** [comment]
Yes, it's definitely the `path = ""`. The reason for this is that the `path` is computed and generated according to the project `name` on GitLab side if it's empty, as described in the docs [here](https://docs.gitlab.com/ee/api/projects.html#create-project:~:text=Generated%20based%20on%20name%20if%20not%20provided). Now, the thing with setting it manually to `""` is that after a `terraform apply` the `path` in the terraform state is correctly set to that generated path from GitLab and a subsequent `terraform apply / plan /...` will detect a change `path "<generated-value" => ""` - however, because an empty `path` is not sent to GitLab during a request you'll only see the `name` field appear there. And the `http_url_to_repo` etc. are all marked as changed, because of the customize diff function I mentioned [earlier](https://github.com/gitlabhq/terraform-provider-gitlab/issues/1208#:~:text=There%20is%20a%20CustomizeDiff%20function%20on%20these%20particular%20attributes%20%2D%20they%20are%20marked%20as%20computed%20if%20either%20the%20namespace_id%20or%20the%20path%20changed.%20Was%20that%20somehow%20the%20case%20(what%20does%20your%20state%20say)%3F).

Bottom line is that setting `path` to `""` never makes any sense and should be detected as an error and rejected by the provider.

I hope this all made sense and clarifies a little bit how this problem came to be 😊

**Comment by thenom** [comment]
Ok, yeh cheers, makes sense about the change of the `name` when `path = ""`.

As a side though, if `PUT {"name": "anything"}` is forbidden via API\Access Token then maybe this should never be attempted as i guess this would be the outcome if someone did change the name.  Obviously not tested this but separate issue if it is.  Cheers again @timofurrer

**Comment by timofurrer** [comment]
@thenom You are welcome, happy to help!

> As a side though, if PUT {"name": "anything"} is forbidden via API\Access Token then maybe this should never be attempted as i guess this would be the outcome if someone did change the name. 

I've just checked and generally changing the `name` using the API should be possible. What kind of a token are you using? Maybe it doesn't have enough permissions?

**Comment by thenom** [comment]
Its a group access token with full permissions.  I guess it's possible that group access tokens have different 'full' rights that other forms of auth and possibly doesnt have this but that's a bit of a minefield in Gitlab (job token, group token, project token, user token, etc...).  I would have thought it would though.
